macaroonbakery/checker.py

```diff
@@ -36,4 +49,388 @@ def __hash__(self):
+        authed = [False]*len(ops)
+        num_authed = 0
+        errors = []
+        for i, op in enumerate(ops):
+            for mindex in self._auth_indexes.get(op, []):
+                _, err = self._check_conditions(ctx, op,
+                                                self._conditions[mindex])
+                if err is not None:
+                    errors.append(err)
+                    continue
+                authed[i] = True
+                num_authed += 1
+                used[mindex] = True
+                # Use the first authorized macaroon only.
+                break
+            if op == LOGIN_OP and not authed[i] and self._identity is not None:
+                # Allow LOGIN_OP when there's an authenticated user even
+                # when there's no macaroon that specifically authorizes it.
+                authed[i] = True
+        if self._identity is not None:
+            # We've authenticated as a user, so even if the operations didn't
+            # specifically require it, we add the login macaroon
+            # to the macaroons used.
+            # Note that the LOGIN_OP conditions have already been checked
+            # successfully in initOnceFunc so no need to check again.
+            # Note also that there may not be any macaroons if the
+            # identity client decided on an identity even with no
+            # macaroons.
+            for i in self._auth_indexes.get(LOGIN_OP, []):
+                used[i] = True
+        if num_authed == len(ops):
+            # All operations allowed.
+            return authed, used
+        # There are some unauthorized operations.
+        need = []
+        need_index = [0]*(len(ops)-num_authed)
+        for i, ok in enumerate(authed):
+            if not ok:
+                need_index[len(need)] = i
+                need.append(ops[i])
+
+        # Try to authorize the operations
+        # even if we haven't got an authenticated user.
+        oks, caveats = self.parent._authorizer.authorize(
+            ctx, self._identity, need)
+        still_need = []
+        for i, _ in enumerate(need):
+            if i < len(oks) and oks[i]:
+                authed[need_index[i]] = True
+            else:
+                still_need.append(ops[need_index[i]])
+        if len(still_need) == 0 and len(caveats) == 0:
+            # No more ops need to be authenticated and
+            # no caveats to be discharged.
+            return authed, used
+        if self._identity is None and len(self._identity_caveats) > 0:
+            raise DischargeRequiredError(
+                msg='authentication required',
+                ops=[LOGIN_OP],
+                cavs=self._identity_caveats)
+        if caveats is None or len(caveats) == 0:
+            all_errors = []
+            all_errors.extend(self._init_errors)
+            all_errors.extend(errors)
+            err = ''
+            if len(all_errors) > 0:
+                err = ' '.join(all_errors)
```

Comment:
Joining the errors with spaces doesn't seem quite right to me. I haven't yet figured out a decent way to format the errors when there's more than one, so perhaps just choose the first one like the Go version does?

Reply:
done.